pkg/conversion/conversion_rules.go

```diff
@@ -0,0 +1,208 @@
+		rulesForVersion = append(rulesForVersion, cr)
+	}
+
+	return rulesForVersion, kerrors.NewAggregate(errs)
+}
+
+func compileRule(rulePath *field.Path, rule apisv1alpha1.APIConversionRule, schema *structuralschema.Structural) (*CompiledRule, error) {
+	cr := &CompiledRule{
+		FromPath: rule.Field,
+		ToPath:   rule.Destination,
+	}
+
+	ruleField := rule.Field
+	if ruleField[0] == '.' {
+		ruleField = ruleField[1:]
+	}
+	schemaForField, err := schemaIndexByJSONPath(schema, ruleField)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("field"), rule.Field, err.Error())
+	}
+
+	if rule.Transformation == "" {
+		return cr, nil
+	}
+
+	env, err := getCELEnv(schemaForField)
+	if env == nil {
+		err = errors.New("environment was nil")
+	}
+	if err != nil {
+		return nil, field.InternalError(rulePath.Child("field"), fmt.Errorf("error creating CEL environment: %w", err))
+	}
+
+	ast, issues := env.Compile(rule.Transformation)
+	if issues != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error compiling CEL program: %v", issues.Err()))
+	}
+
+	program, err := env.Program(ast,
+		cel.EvalOptions(cel.OptOptimize),
+		cel.OptimizeRegex(library.ExtensionLibRegexOptimizations...),
+		cel.InterruptCheckFrequency(celCheckFrequency),
+	)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error creating CEL program: %v", err))
+	}
+
+	cr.Program = program
+
+	return cr, nil
+}
+
+func getCELEnv(structuralSchema *structuralschema.Structural) (*cel.Env, error) {
+	env, err := cel.NewEnv(cel.HomogeneousAggregateLiterals())
+	if err != nil {
+		return nil, fmt.Errorf("error creating CEL environment: %w", err)
+	}
+
+	registry := model.NewRegistry(env)
+
+	// inline local copy of upstream's generateUniqueSelfTypeName()
+	scopedTypeName := fmt.Sprintf("selfType%d", time.Now().Nanosecond())
+
+	ruleTypes, err := model.NewRuleTypes(scopedTypeName, structuralSchema, true, registry)
+	if err != nil {
+		return nil, fmt.Errorf("error creating rule types: %w", err)
+	}
+	if ruleTypes == nil {
+		return nil, nil
+	}
+
+	opts, err := ruleTypes.EnvOptions(env.TypeProvider())
+	if err != nil {
+		return nil, fmt.Errorf("error getting CEL environment options: %w", err)
+	}
+
+	root, ok := ruleTypes.FindDeclType(scopedTypeName)
+	if !ok {
+		rootDecl := model.SchemaDeclType(structuralSchema, true)
+		if rootDecl == nil {
+			return nil, fmt.Errorf("unable to find CEL decl type for %s", structuralSchema.Type)
+		}
+		root = rootDecl.MaybeAssignTypeName(scopedTypeName)
+	}
+
+	var propDecls []*expr.Decl
+	propDecls = append(propDecls, decls.NewVar("self", root.ExprType()))
+
+	opts = append(opts, cel.Declarations(propDecls...), cel.HomogeneousAggregateLiterals())
+	opts = append(opts, library.ExtensionLibs...)
+	return env.Extend(opts...)
+}
+
+func schemaIndexByJSONPath(s *structuralschema.Structural, path string) (*structuralschema.Structural, error) {
```

Comment:
"index" irritates me. Why not `schemaByJSONPath`?

Reply:
I guess you mean `schema(IndexByJSONPath)`, not `(schemaIndex)ByJSONPath` (what I read)

Reply:
Renaming to `getStructuralSchemaForField`